File: myTargetDemo/myTargetDemo/Views/VideoPlayerView.m
//
//  VideoPlayerView.m
//  myTargetDemo
//
//  Created by Andrey Seredkin on 18.01.17.
//  Copyright © 2017 Mail.ru Group. All rights reserved.
//

#import "VideoPlayerView.h"
#import <AVFoundation/AVFoundation.h>

static void *VideoPlayerViewStatusObservationContext = &VideoPlayerViewStatusObservationContext;
static void *VideoPlayerViewRateObservationContext = &VideoPlayerViewRateObservationContext;
static void *VideoPlayerViewExternalPlaybackActiveObservationContext = &VideoPlayerViewExternalPlaybackActiveObservationContext;

@implementation VideoPlayerView
{
	AVPlayer *_player;
	AVPlayerItem *_playerItem;

	AVURLAsset *_asset;
	BOOL _isPaused;
	NSURL * _url;
	float _volume;
	NSTimeInterval _position;
}

- (instancetype)initWithFrame:(CGRect)frame
{
	self = [super initWithFrame:frame];
	if (self)
	{
		_volume = 1;
		_isPaused = NO;
		_position = 0;
	}
	return self;
}

- (void)dealloc
{
	[self stop];
	NSLog(@"VideoPlayerView - dealloc");
}

#pragma mark - delegates

- (void)delegateOnVideoStartedWithUrl:(NSURL *)url
{
	if (!_delegate) return;
	if (![_delegate respondsToSelector:@selector(onVideoStarted:)])return;
	[_delegate onVideoStarted:url];
}

- (void)delegateOnVideoFinishedSuccess
{
	if (!_delegate) return;
	if (![_delegate respondsToSelector:@selector(onVideoFinishedSuccess)])return;
	[_delegate onVideoFinishedSuccess];
}

- (void)delegateOnVideoFinishedWithError:(NSString *)error
{
	[self stop];
	if (!_delegate) return;
	if (![_delegate respondsToSelector:@selector(onVideoFinishedWithError:)])return;
	[_delegate onVideoFinishedWithError:[NSString stringWithFormat:@"VideoPlayerView: %@", error]];
}

#pragma mark - getters

- (NSTimeInterval)currentTime
{
	return _player ? CMTimeGetSeconds([_player currentTime]) : 0.0;
}

- (float)volume
{
	return _player ? _player.volume : _volume;
}

#pragma mark - setters

- (void)setCurrentTime:(NSTimeInterval)currentTime
{
	[_player seekToTime:CMTimeMakeWithSeconds(currentTime, 1)];
}

- (void)setVolume:(float)volume
{
	_volume = volume;
	if (_player)
	{
		_player.volume = volume;
	}
}

#pragma mark - AV player stuff

+ (Class)layerClass
{
	return [AVPlayerLayer class];
}

- (AVPlayer *)player
{
	return [(AVPlayerLayer *) [self layer] player];
}

- (void)setPlayer:(AVPlayer *)player
{
	[(AVPlayerLayer *) [self layer] setPlayer:player];
}

// Specifies how the video is displayed within a player layer’s bounds. AVLayerVideoGravityResizeAspect is default
- (void)setVideoFillMode:(NSString *)fillMode
{
	AVPlayerLayer *playerLayer = (AVPlayerLayer *) [self layer];
	playerLayer.videoGravity = fillMode;
}

#pragma mark - playback

- (void)startWithUrl:(NSURL *)url position:(NSTimeInterval)position
{
	_position = position;
	if (_player && _isPaused && url && _url && [url.absoluteString isEqualToString:_url.absoluteString])
	{
		[_player play];
		return;
	}
	_isPaused = NO;
	_url = url;
	_asset = [AVURLAsset URLAssetWithURL:url options:nil];
	NSArray *requestedKeys = @[@"playable"];

	// Tells the asset to load the values of any of the specified keys that are not already loaded.
	[_asset loadValuesAsynchronouslyForKeys:requestedKeys completionHandler:^
	{
		dispatch_async(dispatch_get_main_queue(),^
		{
			[self prepareToPlayAsset:_asset withKeys:requestedKeys];
		});
	}];
}

- (void)pause
{
	if (!_player) return;
	[_player pause];
	_isPaused = YES;
}

- (void)resume
{
	if (!_player) return;
	_isPaused = NO;
	[_player play];
}

- (void)stop
{
	[_player pause];
	[self deletePlayerItem];
	[self deletePlayer];
}

- (void)finish
{
	[self stop];
	[self delegateOnVideoFinishedSuccess];
}

- (void)prepareToPlayAsset:(AVURLAsset *)asset withKeys:(NSArray *)requestedKeys
{
	/* Make sure that the value of each key has loaded successfully. */
	for (NSString *thisKey in requestedKeys)
	{
		NSError *error = nil;
		AVKeyValueStatus keyStatus = [asset statusOfValueForKey:thisKey error:&error];
		if (keyStatus == AVKeyValueStatusFailed)
		{
			[self delegateOnVideoFinishedWithError:@"Item cannot be played, status failed"];
			return;
		}
		/* If you are also implementing -[AVAsset cancelLoading], add your code here to bail out properly in the case of cancellation. */
	}

	if (!asset.playable)
	{
		[self delegateOnVideoFinishedWithError:@"Item cannot be played"];
		return;
	}

	[self createPlayerItemForAsset:asset];

	[self deletePlayer];
	[self createPlayer];

	[_player seekToTime:CMTimeMakeWithSeconds(_position, 1)];
	[_player play];
}

- (void)createPlayerItemForAsset:(AVURLAsset *)asset
{
	[self deletePlayerItem];
	_playerItem = [AVPlayerItem playerItemWithAsset:asset];

	[_playerItem addObserver:self
				  forKeyPath:@"status"
					 options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
					 context:VideoPlayerViewStatusObservationContext];

	[[NSNotificationCenter defaultCenter] addObserver:self
											 selector:@selector(playerItemDidReachEnd:)
												 name:AVPlayerItemDidPlayToEndTimeNotification
											   object:_playerItem];
}

- (void)deletePlayerItem
{
	if (_playerItem)
	{
		[_playerItem removeObserver:self forKeyPath:@"status"];

		[[NSNotificationCenter defaultCenter] removeObserver:self
														name:AVPlayerItemDidPlayToEndTimeNotification
													  object:_playerItem];
		_playerItem = nil;
	}
}

- (void)createPlayer
{
	[self deletePlayer];

	_player = [AVPlayer playerWithPlayerItem:_playerItem];
	_player.volume = _volume;

	[self setPlayer:_player];

	[_player addObserver:self
			  forKeyPath:@"rate"
				 options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
				 context:VideoPlayerViewRateObservationContext];

	[_player addObserver:self
			  forKeyPath:@"externalPlaybackActive"
				 options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
				 context:VideoPlayerViewExternalPlaybackActiveObservationContext];
}

- (void)deletePlayer
{
	[self setPlayer:nil];
	if (_player)
	{
		[_player removeObserver:self forKeyPath:@"rate"];
		[_player removeObserver:self forKeyPath:@"externalPlaybackActive"];
		[_player pause];
		_player = nil;
	}
}

#pragma mark - Observers

- (void)playerItemDidReachEnd:(NSNotification *)notification
{
	if (notification.object != _playerItem) return;
	[self stop];
	[self delegateOnVideoFinishedSuccess];
}

- (void)observeValueForKeyPath:(NSString *)path
					  ofObject:(id)object
						change:(NSDictionary *)change
					   context:(void *)context
{
	/* AVPlayerItem "status" property value observer. */
	if (context == VideoPlayerViewStatusObservationContext)
	{
		AVPlayerItemStatus status = [[change objectForKey:NSKeyValueChangeNewKey] integerValue];
		[self playerItemStatusChanged:status];
	}
	else if (context == VideoPlayerViewRateObservationContext)
	{
		float rate = [[change objectForKey:NSKeyValueChangeNewKey] floatValue];
		[self playerRateChanged:rate];
	}
	else if (context == VideoPlayerViewExternalPlaybackActiveObservationContext)
	{
		BOOL externalPlaybackActive = [[change objectForKey:NSKeyValueChangeNewKey] boolValue];
		[self externalPlaybackActiveChanged:externalPlaybackActive];
	}
	else
	{
		[super observeValueForKeyPath:path ofObject:object change:change context:context];
	}
}

- (void)playerItemStatusChanged:(AVPlayerItemStatus)status
{
	if (status == AVPlayerItemStatusUnknown)
	{
		NSLog(@"Player item status ---> Unknown");
	}
	else if (status == AVPlayerItemStatusReadyToPlay)
	{
		_duration = [self durationForPlayerItem:_playerItem];
		if (_duration == 0)
		{
			[self delegateOnVideoFinishedWithError:@"Player item duration = 0"];
		}
		else
		{
			// Success
			[self delegateOnVideoStartedWithUrl:_url];
		}
		NSLog(@"Player item status ---> Ready, duration = %.2f", _duration);
	}
	else if (status == AVPlayerItemStatusFailed)
	{
		NSLog(@"Player item status ---> Failed");
		[self delegateOnVideoFinishedWithError:@"Player item status is failed"];
	}
}

- (void)playerRateChanged:(float)rate
{
	NSLog(@"Player rate ---> %.2f", rate);
}

- (void)externalPlaybackActiveChanged:(BOOL)externalPlaybackActive
{
	NSLog(@"AirPlay ---> %@", externalPlaybackActive ? @"YES" : @"NO");
}

#pragma mark - helpers

- (NSTimeInterval)durationForPlayerItem:(AVPlayerItem *)playerItem
{
	if (playerItem && playerItem.status == AVPlayerItemStatusReadyToPlay)
	{
		CMTime playerDuration = [playerItem duration];
		if (CMTIME_IS_INVALID(playerDuration))
		{
			return 0;
		}
		double duration = CMTimeGetSeconds(playerDuration);
		if (isfinite(duration))
		{
			return duration;
		}
		else
		{
			return 0;
		}
	}
	else
	{
		return 0;
	}
}

@end

File: myTargetDemoSwift/myTargetDemo/Views/VideoPlayerView.swift
//
//  VideoPlayerView.swift
//  myTargetDemo
//
//  Created by Andrey Seredkin on 01/08/2019.
//  Copyright © 2019 Mail.Ru Group. All rights reserved.
//

import UIKit
import AVFoundation

protocol VideoPlayerViewDelegate: AnyObject {
	func onVideoStarted(url: URL)
	func onVideoComplete()
	func onVideoFinished(error: String)
}

final class VideoPlayerView: UIView {
	weak var delegate: VideoPlayerViewDelegate?

	private var position: TimeInterval = 0.0
	private var isPaused = false

	private var _currentTime: TimeInterval = 0.0
	var currentTime: TimeInterval {
		get {
			guard let player = player else {
                return _currentTime
            }

			return CMTimeGetSeconds(player.currentTime())
		}
		set {
			_currentTime = newValue
			guard let player = player else {
                return
            }

			player.seek(to: CMTimeMakeWithSeconds(_currentTime, preferredTimescale: 1))
		}
	}

	private var _volume: Float = 1.0
	var volume: Float {
		get {
			guard let player = player else {
                return _volume
            }

			return player.volume
		}
		set {
			_volume = newValue
			guard let player = player else {
                return
            }

			player.volume = _volume
		}
	}

	private var url: URL?
	private var asset: AVURLAsset?
	private var playerItem: AVPlayerItem?

	private var player: AVPlayer? {
		get {
			guard let layer = layer as? AVPlayerLayer else {
                return nil
            }

			return layer.player
		}
		set {
			guard let layer = layer as? AVPlayerLayer else {
                return
            }

			layer.player = newValue
		}
	}

	private static var playerItemContext = 0
	private let requiredAssetKeys = ["playable"]

	private var playerItemDuration: TimeInterval {
        guard let playerItem = playerItem, playerItem.status == .readyToPlay, CMTIME_IS_VALID(playerItem.duration) else {
            return 0.0
        }

        let duration = CMTimeGetSeconds(playerItem.duration)
        return duration.isFinite ? duration : 0.0
	}

	deinit {
		stop()
	}

	override final class var layerClass: AnyClass {
		return AVPlayerLayer.self
	}

// MARK: - public

	func start(with url: URL, position: TimeInterval) {
		self.position = position
		if let player = player, isPaused, let internalURL = self.url, internalURL.absoluteString == url.absoluteString {
			player.play()
            delegateOnVideoStarted(url: url)
			return
		}

		self.url = url
		isPaused = false
		asset = AVURLAsset(url: url)
		guard let asset = asset else {
            return
        }

		asset.loadValuesAsynchronously(forKeys: requiredAssetKeys) {
			DispatchQueue.main.async {
				self.prepareToPlay()
			}
		}
	}

	func pause() {
		guard let player = player else {
            return
        }

		player.pause()
		isPaused = true
	}

	func resume() {
		guard let player = player else {
            return
        }

		player.play()
		isPaused = false
	}

	func stop() {
		player?.pause()
		deletePlayerItem()
		deletePlayer()
	}

	func finish() {
		stop()
		delegateOnVideoComplete()
	}

// MARK: - private

	private func prepareToPlay() {
		guard let asset = asset else {
            return
        }

		requiredAssetKeys.forEach { (key: String) in
			var error: NSError?
			let keyStatus = asset.statusOfValue(forKey: key, error: &error)
			if keyStatus == .failed {
				stop()
				delegateOnVideoFinished(error: "Item cannot be played, status failed")
			}
		}

		if !asset.isPlayable {
			stop()
			delegateOnVideoFinished(error: "Item cannot be played")
			return
		}

		deletePlayerItem()
		createPlayerItem()

		deletePlayer()
		createPlayer()

		guard let player = player else {
            return
        }

		player.seek(to: CMTimeMakeWithSeconds(position, preferredTimescale: 1))
		player.play()
	}

	private func createPlayerItem() {
		guard let asset = asset else {
            return
        }

		playerItem = AVPlayerItem(asset: asset)
		guard let playerItem = playerItem else {
            return
        }

		playerItem.addObserver(self, forKeyPath: #keyPath(AVPlayerItem.status), options: [.initial, .new], context: &VideoPlayerView.playerItemContext)
		NotificationCenter.default.addObserver(self, selector: #selector(playerItemDidReachEnd), name: .AVPlayerItemDidPlayToEndTime, object: playerItem)
	}

	private func deletePlayerItem() {
		guard let playerItem = playerItem else {
            return
        }

		playerItem.removeObserver(self, forKeyPath: #keyPath(AVPlayerItem.status))
		NotificationCenter.default.removeObserver(self, name: .AVPlayerItemDidPlayToEndTime, object: playerItem)
		self.playerItem = nil
	}

	private func createPlayer() {
		guard let playerItem = playerItem else {
            return
        }

		player = AVPlayer(playerItem: playerItem)
		guard let player = player else {
            return
        }
		player.volume = volume
	}

	private func deletePlayer() {
		guard let player = player else {
            return
        }

		player.pause()
		self.player = nil
	}

// MARK: - observers

	@objc private func playerItemDidReachEnd(notification: Notification) {
		guard let object = notification.object as? AVPlayerItem, let playerItem = playerItem, object == playerItem else {
            return
        }

		stop()
		delegateOnVideoComplete()
	}

    // swiftlint:disable:next block_based_kvo
	override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey: Any]?, context: UnsafeMutableRawPointer?) {
		guard context == &VideoPlayerView.playerItemContext else {
			super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
			return
		}
		guard keyPath == #keyPath(AVPlayerItem.status) else {
            return
        }

        let statusNumber = change?[.newKey] as? NSNumber
        let status = statusNumber.flatMap {
            AVPlayerItem.Status(rawValue: $0.intValue)
        } ?? .unknown

		playerItemStatusChanged(status)
	}

    private func playerItemStatusChanged(_ status: AVPlayerItem.Status) {
        switch status {
        case .readyToPlay:
            let duration = playerItemDuration
            print("Player item status ---> Ready, duration = \(duration)")
            if duration == 0 {
                stop()
                delegateOnVideoFinished(error: "Player item duration = 0")
            } else {
                // Success
                guard let url = url else {
                    return
                }
                delegateOnVideoStarted(url: url)
            }
        case .failed:
            print("Player item status ---> Failed")
            stop()
            delegateOnVideoFinished(error: "Player item status is failed")
        case .unknown:
            print("Player item status ---> Unknown")
        @unknown default:
            break
        }
    }

// MARK: - delegates

	private func delegateOnVideoStarted(url: URL) {
		delegate?.onVideoStarted(url: url)
	}

	private func delegateOnVideoComplete() {
		delegate?.onVideoComplete()
	}

	private func delegateOnVideoFinished(error: String) {
		delegate?.onVideoFinished(error: error)
	}
}
